In [33]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris

In [34]:
df = load_iris()

In [35]:
df

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [36]:
iris1 = pd.DataFrame(df.data, columns = df.feature_names)

In [37]:
y = df.target

In [38]:
from sklearn.ensemble import RandomForestClassifier
rfmodel = RandomForestClassifier(n_estimators=100, criterion='gini', max_depth=4)

In [39]:
rfmodel.fit(iris1, y)

RandomForestClassifier(max_depth=4)

In [40]:
rfmodel.score(iris1, y)

0.9866666666666667

In [41]:
rfmodel2 = RandomForestClassifier(n_estimators=10, criterion='gini', max_depth=4)

In [42]:
rfmodel2.fit(iris1, y)
rfmodel2.score(iris1, y)

0.98

In [43]:
from sklearn.model_selection import cross_val_score
cross_val_score(rfmodel2, iris1, y, cv=5, scoring='accuracy')

array([0.96666667, 0.96666667, 0.93333333, 0.9       , 1.        ])

In [44]:
for i in [1,5,10,20,50,100,200,300]:
    rfmodel2.n_estimators = i
    print(i, cross_val_score(rfmodel2, iris1, y, cv=5, scoring='accuracy') )

1 [0.93333333 0.96666667 0.86666667 0.93333333 1.        ]
5 [0.96666667 0.96666667 0.9        0.96666667 1.        ]
10 [0.96666667 0.96666667 0.9        0.9        1.        ]
20 [0.96666667 0.96666667 0.93333333 0.9        1.        ]
50 [0.96666667 0.96666667 0.9        0.96666667 1.        ]
100 [0.96666667 0.96666667 0.93333333 0.93333333 1.        ]
200 [0.96666667 0.96666667 0.93333333 0.93333333 1.        ]
300 [0.96666667 0.96666667 0.93333333 0.93333333 1.        ]


Welcome to the software demonstration on random forest. Let's go ahead and get started, do our standard imports. All right, so you had a little problem with the iris. Let's deal with that in just a second and get our stuff in.

All right, so what we're going to do is, we're going to use the iris dataset. It might be possibly capitalized. Let's give it a shot here. It's probably the load iris. There we go.

All right, so once again, we'll get this array. And we'll have to take a look at the data types to get it. We see we've got a target. We've got the target names. All right, so we're going to have to do a little modification of that.

And now, we've got down here the feature names. So these will be our column names. So we'll have to remember that as we create our DataFrame.

So we'll just call this iris data. That may seem a little crazy to have data data. But remember, I put that in a data object.

And as I scroll up here, what you may not realize is, there's a key in the dictionary called data. So for our DataFrame, that's what we're going to do. We'll need an actual DataFrame. So there we go.

And of course, the columns are the feature names. So let's just double check and make sure we're calling that correctly. Yep, feature names, and we can even copy it directly out of there. Let's take a look.

So this is our classic iris dataset. We'll also needs some targets here. We typically call that y. Let's take a look at what that's looking like.

All right, so that's already performed as integers for us. So we've got a multiclass classification, something that's slightly complex. The iris dataset isn't too bad. But we've got a multiclass classification. So we want to take our first attempt at modeling this.

So our first attempt is the random forest. Random forest is located in the ensemble part of the package. Now, you'll notice there's a couple of options. Since we're doing a classification, we'll obviously want the RandomForestClassifier. If we were doing the regressor, we would choose the latter.

So let's start with the RandomForestClassifier. And of course, I like to do my model names by initials. But let's just call it our rf_model.

Now, in the past couple of videos, I'd like to pull up the documentation just to refresh, because there is so much in this package. So let's go ahead and take a look at the random forest.

We're doing the RandomForestClassifier. So we can see now we're dealing with a quite complex class that has a number of things that we need to control here at the beginning. One of the most important is the number of estimators.

You'll notice the default is 100. So that means if you add no extra information and just use the default with the rounded braces, you'll get a random forest with 100 trees. That's the number of estimators.

We see that the default criterion is gini. Now, we've studied both gini and entropy. And so we can switch back and forth between them. We'll just keep in the back of our mind that the default is using gini for information gain.

Now we start to see all the things that we control for our partition trees. And of course, each tree will be subject to that. And that includes the max depth, the number of samples required for a split, the number of samples for a leaf, and of course, things like weight fractions and the number of features.

Now, you'll notice these limits are set quite low. And of course, for a random forest, we're not concerned with the individual tree being overfit. In fact, we don't care if they are greatly overfit. Because what's important for us is the overall ensemble. And the overfit trees selected from random data will actually help us by being less correlated to each other.

That may seem counterintuitive, but it's actually one of the powers of the random classifier. So we've seen kind of what we need to do. So let's make some initial guesses on what our classifier might do.

We've seen that the number of estimators is default 100, but let's just put that in. And let's put in our criteria. Now, let's see if what happens-- we'll start with a max depth. And let's just try something like 4.

And again, these are guesses. And we're going to have to tune all of these. But we know that from our work on partition trees, that the iris data set doesn't have a lot. So we probably don't want to do too much depth.

So we've got kind of our initial setup, our initial guess. We've put in some parameters, not all. And again, each time you do this, you're going to have to make a decision on what parameters you want to check. And we'll start to see that each of these takes more and more time to fit.

So we've got all this set up. We're going to create our model. Now it's time to go ahead and fit our model. Let's remember what we actually called it, our iris_data.

So that didn't actually take too long. And let's see what we have when we fit this model. So you can see we've got a number of issues, or not issues, but a number of things we can take a look at.

We see the estimator parameters. We can see the individual trees. You'll notice we have feature importances.

We may not have talked about feature importances, but it's handy to know in random forests. And it turns out the feature importance is related to how many times that feature was used to make a split. So this can be handy for helping us to interpret these forest classifiers.

So while we often rely on linear models for feature importance, remember, you can do it for random forests too, it's just less clear. Because it's not about the relationship to the output. It's more about how often that parameter was used to make a split.

We can see all our different variables. And what we're looking for is the score. Now, you may wonder, what is this oob_score? What this is going to be is our out-of-bagging score.

Because random forest takes a bagging approach, what that means is the samples that were not used to build the trees were actually run through as a pseudo test set. So this can be used as a replacement for your basically out-of-fold predictions. Although, because there is some bias involved, it's not always the best choice.

We're still going to go with our cross-validation. But be aware that this is available for your random forest. What we're really looking for is the score.

Let's take a look at that score. And we'll need to score our data. And so, we see this is actually an accuracy score, and we've got a fairly accurate model. So let's see what can happen as we start to change some of these parameters.

So instead of having 100 estimators, let's just crank this down to 10. So I'm just going to redo my model, set the estimators to 10. Let's fit the model again. And let's take a look at the score.

Now again, this is the biased score. But we're just looking at what happens as we vary the model parameters. You'll notice we still get a pretty high score.

But notice something key. We're losing some precision. Not precision of the model, but we're down to getting scores that, instead of out to multiple decimal places, come down to a single decimal place. In fact, if we start doing estimates, you'll see that our scores will become 90, 80, 70, as we crank these up and down. Because the score is actually related.

If we looked at accuracy in particular, it is related to the vote. And because there are 10 estimators in our vote, we would have that as our score. Let's see if we can approximate that with our cross-validation.

For a reminder, we always do cross-validation because we don't want to get biased estimates of our score. And so we need the out-of-fold predictions. Let's just be sure we're using five folds. And let's see if we can get our accuracy in here.

So you can see, what we're getting is some of the folds are extremely accurate, while some of them vary from 96% down to 93%. Let's see what happens as we start to change some of the parameters. So let's see if we can set the number of estimators in our random forest model. So it looks like we can. So let's see if we can use our previous method of iterating through and watching how the scores change.

I'm going to do something a little bit different this time. I'm going to provide a list. So we're just going to select each item out of that list and see what happens to the score. Close up our parentheses and watch what happens.

Now, you'll notice that the scores didn't change that much. So what's happening in the random forest, and is unique about the random forests, is we don't suffer so much from overfitting. We may have to tune to get the best model. But even as we crank up the number of estimators, the overfitting really doesn't become a problem.

You may be concerned about this last value dropping from 96 to 93. It's actually because the data set is so small, that represents a single data point. So while that may be an issue for us, in general, as you crank up the number of estimators in a random forest, you really don't suffer much.

So it's to your advantage to use the largest number of estimators that you can. This is one of the unique features of random forests. However, it doesn't apply to any of the other parameters, only to the number of estimators.

Now, if you review all the things that we had, we could go through this loop multiple times. We could be here all day trying to vary these parameters. And in future lectures, we'll talk about how to properly tune various models with multiple parameters.

So we won't do that here. But hopefully this starts to give you an idea of what to do for the random forest. But let's take a look at what happens when we vary some other parameters, just to get an understanding for the random forest and become comfortable with using it.

Let's see what happens when we switch to the entropy loss or the entropy information gain. So we're going to create a new model. I'll just copy this down here. We'll maybe denote it by entropy. And we'll switch this, go back to our default of 100.

And let's see what happens when we fit this model. Let's actually do a cross-validation score to compare. You can see there really isn't much change between entropy and gini.

So this is a key takeaway. That while there might be a specific problem with entropy or gini, it really doesn't matter which one you use. And it's really important to figure out which one gives you the better result. Once again, there's no hard-and-fast rule.

Some of the other things we want to take a look at is the maximum depth. But before we do that, let's take a look at the documentation to see if maybe there's some guidance about this. So if we look at the max depth here, we see it's the max depth of the tree. And if we leave it blank, it means everything is expanded until all the other criteria are met. That's essentially what this is saying.

Now, we intentionally limited this to 4. So let's see what happens when we turn it off. There's one other thing I want to take a look at while we're here. And that's the max features. Now, we didn't input anything for max features. And you'll notice there's various options here. And we get a little blurb about each one.

So if we actually provide an integer, it's going to use that integer features for split. So if I say the max features is 10, what the random forest is going to do is it's going to look at 10 features to consider the split. If we provide a float, that will be a percentage of the features. If we provide auto, that means we're going to use the square root of the number of features.

This is kind of your handy rule of thumb for forests for how many max features to use. In general, it should be the square root of N. But again, each problem is different. So this is something that we can experiment.

Again, we may not go through all the iterations at this point. But I'm trying to highlight the important ones that you can spend your time on that are going to give you the most benefit rather than playing with every single one and trying to find the max improvement. So in general, what you're going to want to look for is the max features, the max depth, and possibly switching between your different criteria. Again, the estimators, you really don't have to play with that. You set it to a large number, really as large as you can tolerate as terms of time.

Hopefully, you saw in the video that as we went to larger and larger numbers, it does slow down. But the other important thing is, there is hidden in here a number of jobs. So if you have a multiprocessor computer, and most people do, don't forget to take advantage of this to allow your multiprocessing to help speed up your job. And again, that allows you to get a little bit more accuracy.

So let's take a look and just do how things change as we swap a couple of these things. So we'll keep our entropy model. But now, instead of the max depth 4, we'll just eliminate it, and it'll be as many features as we want. And let's see if we get much of a difference.

So you can see we actually get a little bit of an improvement on this. It's not significant, because we were already pretty high. But this shows you that these parameters do have an impact on your final results.

Now, the iris data set is a fairly simple data set. So it's pretty easy to get good results. But it's also a good way for you to learn your new classifiers.

And remember, we didn't do anything with multiclass. The random forest was able to handle it automatically. That's why I like to use this as my first choice when I'm taking an attempt to make a new model.

The random forest is extremely user friendly, is able to handle multiple classes, and is pretty easy to optimize. We usually, once I've got a good score out of my random forest, then I try and beat it with my more advanced models. That tells me that I'm improving on my score.

If I can't beat the random forest score, that tells me that something's wrong with my advanced models. Random forest kind of stands as our baseline for what we should be able to achieve. Keep that in mind as you go through when you're tuning these models.

This concludes our video on the random forest classifier.